apps/ios/Shared/Views/Chat/ChatView.swift

```diff
@@ -447,17 +456,74 @@ struct ChatView: View {
     }
     
     @ViewBuilder private func chatItemView(_ ci: ChatItem, _ maxWidth: CGFloat) -> some View {
-        if case let .groupRcv(member) = ci.chatDir,
-           case let .group(groupInfo) = chat.chatInfo {
-            let (prevItem, nextItem) = chatModel.getChatItemNeighbors(ci)
-            if ci.memberConnected != nil && nextItem?.memberConnected != nil {
-                // memberConnected events are aggregated at the last chat item in a row of such events, see ChatItemView
+        ChatItemWithMenu(
+            chat: chat,
+            chatItem: ci,
+            maxWidth: maxWidth,
+            composeState: $composeState,
+            selectedMember: $selectedMember,
+            chatView: self
+        )
+    }
+
+    private struct ChatItemWithMenu: View {
+        @EnvironmentObject var m: ChatModel
+        @Environment(\.colorScheme) var colorScheme
+        @ObservedObject var chat: Chat
+        var chatItem: ChatItem
+        var maxWidth: CGFloat
+        @Binding var composeState: ComposeState
+        @Binding var selectedMember: GMember?
+        var chatView: ChatView
+
+        @State private var deletingItem: ChatItem? = nil
+        @State private var showDeleteMessage = false
+        @State private var deletingItems: [Int64] = []
+        @State private var showDeleteMessages = false
+        @State private var revealed = false
+        @State private var showChatItemInfoSheet: Bool = false
+        @State private var chatItemInfo: ChatItemInfo?
+
+        @State private var allowMenu: Bool = true
+
+        @State private var audioPlayer: AudioPlayer?
+        @State private var playbackState: VoiceMessagePlaybackState = .noPlayback
+        @State private var playbackTime: TimeInterval?
+
+        var body: some View {
+            let (currIndex, nextItem) = m.getNextChatItem(chatItem)
+            let ciCategory = chatItem.mergeCategory
+            if (ciCategory != nil && ciCategory == nextItem?.mergeCategory) {
+                // memberConnected events and deleted items are aggregated at the last chat item in a row, see ChatItemView
                 ZStack {} // scroll doesn't work if it's EmptyView()
             } else {
-                if prevItem == nil || showMemberImage(member, prevItem) {
+                let (prevHidden, prevItem) = m.getPrevShownChatItem(currIndex, ciCategory)
+                let range = itemsRange(currIndex, prevHidden)
+                if revealed, let range = range {
+                    let items = Array(zip(Array(range), m.reversedChatItems[range]))
+                    ForEach(items, id: \.1.viewId) { (i, ci) in
```

Comment:
this loop renders items from the current merged group if the current item is "revealed" or "expanded".